### PR TITLE
feat: Add listener for initial state on flame_bloc

### PR DESCRIPTION
**Pull request by luanpotter**
# Description

While analyzing the tests for flame_bloc, I stumbled upon an interesting conundrum.
It seems that the `initialState` provided does not trigger a state change callback through `onNewState`. While that is in line with bloc, I think for most game use cases it is important to keep track of the "last" state of stuff - and in fact most of our existing examples do use the library in that way, to keep track of the last (i.e., the current) state.
However, doing so generically is not simple with the current library because of how the initial state is not provided. While you could initialize your last manually, if there is a dynamic initial state, it would not be trivial to wire it up.
This, therefore, adds a new listener of `onInitialState`, maintaining the behaviour of `onNewState`.

Note: this solves the last unused reference [from this issue](https://github.com/flame-engine/flame/issues/2308) by using it on the new tests.

## Checklist

- [x] I have followed the [Contributor Guide] when preparing my PR.
- [x] I have updated/added tests for ALL new/updated/fixed functionality.
- [x] I have updated/added relevant documentation in `docs` and added dartdoc comments with `///`.
- [x] I have updated/added relevant examples in `examples` or `docs`.

## Breaking Change?

- [ ] Yes, this PR is a breaking change.
- [x] No, this PR is not a breaking change.

[Contributor Guide]: https://github.com/flame-engine/flame/blob/main/CONTRIBUTING.md
[Conventional Commit]: https://conventionalcommits.org
[CHANGELOG]: https://github.com/flame-engine/flame/blob/main/CHANGELOG.md

### CODE REVIEW

File: packages/flame_bloc/lib/src/flame_bloc_listenable.dart [hunk 1]
```diff
@@ -50,6 +50,7 @@ mixin FlameBlocListenable<B extends BlocBase<S>, S> on Component {
     }
     _bloc = bloc;
     _state = bloc.state;
+    onNewState(_state);
```
Comment (luanpotter) [inline packages/flame_bloc/lib/src/flame_bloc_listenable.dart:53]: this is the only production change
the rest is fixing & adding tests

**Comment by KyleKun** [comment]
> However, if we want to avoid the breaking change, the other option is to add a second callback hook, analogous to onNewState, but called onInitialState that of course is called only once.

I prefer this approach, makes more sense than triggering `onNewState` imo.

**Comment by luanpotter** [comment]
I agree, let me try out that approach.

**Review by spydon — APPROVED** [review]

File: packages/flame_bloc/test/src/flame_bloc_listenable_test.dart [hunk 2]
```diff
@@ -78,7 +92,7 @@ void main() {
 
         bloc.makeSad();
         await Future.microtask(() {});
-        expect(component.last, isNull);
+        expect(component.last, PlayerState.alive);
```
Comment (erickzanardo) [inline packages/flame_bloc/test/src/flame_bloc_listenable_test.dart:95]:
```suggestion
        expect(component.last, equals(PlayerState.alive));
```

The same suggestion a couple of line bellow

Comment (luanpotter) [reply packages/flame_bloc/test/src/flame_bloc_listenable_test.dart:95]: what's the difference?

Comment (erickzanardo) [reply packages/flame_bloc/test/src/flame_bloc_listenable_test.dart:95]: Matchers give better error message when a test fails

File: packages/flame_bloc/test/src/flame_bloc_listenable_test.dart [hunk 3]
```diff
@@ -96,7 +110,7 @@ void main() {
 
         bloc.kill();
         await Future.microtask(() {});
-        expect(component.last, isNull);
+        expect(component.last, PlayerState.alive);
```
Comment (luanpotter) [inline packages/flame_bloc/test/src/flame_bloc_listenable_test.dart:113]:
```suggestion
        expect(component.last, equals(PlayerState.alive));
```

**Review by erickzanardo — APPROVED** [review]

**Comment by joaosilva2095** [comment]
Hi everyone 👋

Great work on adding this feature! It is something definitely useful 😊
While checking the changes, I have noticed that the `FlameBlocListener` was not updated with the new `onInitialState` parameter.

Was this the goal? Or should that parameter be added and passed to the `FlameBlocListenable`?

**Comment by erickzanardo** [comment]
Good callout @joaosilva2095 that totally escaped our attention, here is the PR addressing it: https://github.com/flame-engine/flame/pull/2565

**Pull request merged**